pkg/pool-stable-phantom/test/StablePhantomPool.test.ts

```diff
@@ -641,17 +640,168 @@ describe('StablePhantomPool', () => {
+
+            expect(result.bptIn).to.equal(previousBptBalance);
+            expect(result.amountsOut).to.be.lteWithError(expectedAmountsOut, 0.00001);
+          }
+
+          it('can tell how many tokens it will give in return', async () => {
+            await multiExitGivenIn();
+          });
+
+          it('reverts if paused', async () => {
+            await pool.pause();
+
+            const bptIn = previousBptBalance.div(2);
+            await expect(pool.multiExitGivenIn({ from: lp, bptIn })).to.be.revertedWith('PAUSED');
+          });
+        }
+      });
+
+      describe('exit BPT in for exact tokens out', () => {
+        context('not in recovery mode', () => {
+          itExitsBptInForExactTokensOutProperly();
+        });
+
+        context('in recovery mode', () => {
+          sharedBeforeEach('enable recovery mode', async () => {
+            await pool.enableRecoveryMode(admin);
+          });
+
+          itExitsBptInForExactTokensOutProperly();
+        });
+
+        function itExitsBptInForExactTokensOutProperly() {
+          it('grants exact tokens for bpt', async () => {
+            // Request half of the token balances
+            const amountsOut = initialBalances.map((balance) => bn(balance).div(2));
+
+            const expectedBptIn = previousBptBalance.div(2);
+            const maximumBptIn = pct(expectedBptIn, 1.01);
+
+            const result = await pool.exitGivenOut({ from: lp, amountsOut, maximumBptIn });
+
+            // Token balances should been reduced as requested
+            expect(result.amountsOut).to.deep.equal(amountsOut);
+
+            // BPT balance should have been reduced by half because we are returning half of the tokens
+            expect(await pool.balanceOf(lp)).to.be.equalWithError(previousBptBalance.div(2), 0.001);
+          });
+
+          async function exitGivenOut(): Promise<void> {
+            const amountsOut = initialBalances.map((balance) => bn(balance).div(2));
+            const expectedBptIn = previousBptBalance.div(2);
+            const maximumBptIn = pct(expectedBptIn, 1.01);
+
+            const result = await pool.queryExitGivenOut({ amountsOut, maximumBptIn });
+
+            expect(result.amountsOut).to.deep.equal(amountsOut);
+            expect(result.bptIn).to.be.equalWithError(previousBptBalance.div(2), 0.001);
+          }
+
+          it('fails if more BTP needed', async () => {
+            // Call should fail because we are requesting a max amount lower than the actual needed
+            const amountsOut = initialBalances;
+            const maximumBptIn = previousBptBalance.div(2);
+
+            await expect(pool.exitGivenOut({ from: lp, amountsOut, maximumBptIn })).to.be.revertedWith(
+              'BPT_IN_MAX_AMOUNT'
+            );
+          });
+
+          it('can tell how much BPT it will have to receive', async () => {
+            await exitGivenOut();
+          });
```

Comment:
```suggestion
          it('fails if more BTP needed', async () => {
            // Call should fail because we are requesting a max amount lower than the actual needed
            const amountsOut = initialBalances;
            const maximumBptIn = previousBptBalance.div(2);

            await expect(pool.exitGivenOut({ from: lp, amountsOut, maximumBptIn })).to.be.revertedWith(
              'BPT_IN_MAX_AMOUNT'
            );
          });

          it('can tell how much BPT it will have to receive', async () => {
            const amountsOut = initialBalances.map((balance) => bn(balance).div(2));
            const expectedBptIn = previousBptBalance.div(2);
            const maximumBptIn = pct(expectedBptIn, 1.01);

            const result = await pool.queryExitGivenOut({ amountsOut, maximumBptIn });

            expect(result.amountsOut).to.deep.equal(amountsOut);
            expect(result.bptIn).to.be.equalWithError(previousBptBalance.div(2), 0.001);
          });
```

Reply:
ditto